doc/adr/0010-reducing-snapshot-size.md

```diff
@@ -0,0 +1,343 @@
+    the `AggregatedInfo` is the same as `PartialInfo`. Missing values has the same meaning as in `PartialInfo`.  
+
+```json
+{
+  "ordinal": 1500,
+  "info": {
+    "partial": {
+      "balances": {},
+      "lastTransactionRefs": {},
+      "lastSnapshotHashes": {}
+    },
+    "aggregated": {
+      "since": 1001,
+      "balances": {},
+      "lastTransactionRefs": {},
+      "lastSnapshotHashes": {}
+    }
+  }
+}
+```
+
+### Aggregation schemes
+
+One of the advantages of this representation are flexible aggregation schemes.
+
+Simple scheme: 
+* if ordinal is divisible by 1000 do aggregate since 0 (genesis)
+* otherwise do aggregate since last ordinal divisible by 1000
+```mermaid
+graph RL;
+    A(ordinal = 1999, since = 1001)
+    B(ordinal = 1000, since = 0)
+    A-->B;
+```
+
+More complicated scheme:
+* if ordinal is divisible by 1000 do aggregate since 0 (genesis)
+* if ordinal is divisible by 100 do aggregate since last ordinal divisible by 1000
+* if ordinal is divisible by 10 do aggregate since last ordinal divisible by 100
+* otherwise, do aggregate since last ordinal divisible by 10
+```mermaid
+graph RL;
+    A(ordinal = 1999, since = 1991)
+    B(ordinal = 1990, since = 1901)
+    C(ordinal = 1900, since = 1001)
+    D(ordinal = 1000, since = 0)
+    A-->B;
+    B-->C;
+    C-->D;
+```
+
+### Aggregated view resolution
+
+Algorithm for resolving an aggregated view of all data - from the genesis to the latest snapshot:
+1. Variable initialization:
+    * `snapshot := <latest-snapshot>`
+    * `acc := snapshot.info.aggregated`
+2. Loop:
+    * if `snapshot.info.aggregated.since` is equal to `0` return `acc`
+    * otherwise, get the snapshot with ordinal `snapshot.info.aggregated.since` and assign it to `snapshot`
+    * combine `acc` with `snapshot.info.aggregated` and assign it to `acc`
+
+
+### Example of first 2000 snapshots
+
+This example uses a **Simple scheme** presented in section [Aggregation schemes](#aggregation-schemes). 
+
+Simplifications:
+* all snapshots that are not listed below didn't contain any transactions
+* rewards were 0 for all snapshots
+* some fields were omitted for clarity
+
+Snapshots:
+- [Snapshot 0 (genesis)](#snapshot-0)
+- [Snapshot 1](#snapshot-1)
+- [Snapshot 999](#snapshot-999)
+- [Snapshot 1000](#snapshot-1000)
+- [Snapshot 1001](#snapshot-1001)
+- [Snapshot 1999](#snapshot-1999)
+- [Snapshot 2000](#snapshot-2000)
+
+#### Snapshot 0
+
+```json
+{
+  "ordinal": 0,
+  "blocks": [],
+  "info": {
+    "partial": {
+      "balances": {
+        "A": 20,
+        "B": 20
+      }
+    },
+    "aggregated": {
+      "since": 0,
+      "balances": {
+        "A": 20,
+        "B": 20
+      }
+    }
+  }
+}
+```
+
+#### Snapshot 1
+
+```json
+{
+  "ordinal": 1,
+  "blocks": [
+    {
+      "transactions": [
+        {
+          "source": "B",
+          "destination": "C",
+          "amount": 10
+        }
+      ]
+    }
+  ],
+  "info": {
+    "partial": {
```

Comment:
I'm starting to think that having `partial` as well as `aggregated` gives us nothing. It is isomorphic to the blocks added in this snapshot and gives nothing during download/redownload on L1.

Reply:
Please explain how `aggregated` is isomorphic with blocks added in this snapshot? 

`partial` is also not isomorphic btw. You can have two transactions for the same address that are updating balances with values `+1`  and `+1` or a single transaction with `+2`, or `+3 and -1`, etc... Isomorphism is injective (1 to 1 function, never 1 to many).

That being said, `partial` is a simplified view of the same data that is already in transactions and state channel snapshots, however, I think it's worth having it since clients processing a stream of global snapshots will have a clear view of what values have been updated in every single snapshot without having to apply those blocks and transactions manually.

Reply:
Exactly, the idea of partial data is to be able to quickly determine what has changed in a single snapshot (hence block explorer can use that).

Reply:
That's what I meant by saying isomorphic (so I used it wrongly), that by folding transactions you will get exactly what partial is.

Reply:
Whereas aggregated gives some added value, because we don't have these transactions in this snapshot. So partial is kind of duplication of information. But if we think it's a worthy optimization then let's have it. What bothers me is the ever problematic decoupling between transactions and folded balances from these transactions, they should be the same but programmatically it's not guaranteed.

Reply:
@TheMMaciek yes, but you have to have previous balances stored locally to apply the transactions, so clients are getting more complicated.

Reply:
yes, but the opposite direction does not hold